flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1080,18 +1080,22 @@ else if (current == ExecutionState.RUNNING) {
 	// ------------------------------------------------------------------------
 
 	@Override
-	public CompletableFuture<PartitionProducerStateResponseHandle> requestPartitionProducerState(
+	public void requestPartitionProducerState(
 			final IntermediateDataSetID intermediateDataSetId,
-			final ResultPartitionID resultPartitionId) {
+			final ResultPartitionID resultPartitionId,
+			Consumer<? super ResponseHandle> responseConsumer) {
+
 		final CompletableFuture<ExecutionState> futurePartitionState =
 			partitionProducerStateChecker.requestPartitionProducerState(
 				jobId,
 				intermediateDataSetId,
 				resultPartitionId);
-		final CompletableFuture<PartitionProducerStateResponseHandle> result =
-			futurePartitionState.handleAsync(PartitionProducerStateResponseHandle::new, executor);
-		FutureUtils.assertNoException(result);
-		return result;
+
+		futurePartitionState.whenCompleteAsync(
+			(ExecutionState executionState, Throwable throwable) -> {
+				responseConsumer.accept(new PartitionProducerStateResponseHandle(executionState, throwable));
+			},
+			executor);
```

Comment:
nit: one could also write it

```
FutureUtils.assertNoException(
			futurePartitionState
				.handle(PartitionProducerStateResponseHandle::new)
				.thenAcceptAsync(responseConsumer, executor));
```

Which also ensures that any exception coming out of the consumer will make the process to fail.